MiniApp/Classes/MiniAppDownloader.swift

```diff
@@ -22,52 +18,70 @@ class MiniAppDownloader: NSObject, MiniAppDownloaderProtocol {
-        let configuration = URLSessionConfiguration.default
-        return URLSession(configuration: configuration, delegate: self, delegateQueue: queue)
-    }()
+    init() {
+        self.miniAppStorage = MiniAppStorage()
+        self.manifestDownloader = ManifestDownloader()
+    }
 
-    func download(_ urls: [String], to pathDirectory: URL) {
-        for url in urls {
-            guard let fileDirectory = UrlParser.parseForFileDirectory(with: url),
-                let downloadUrl = URL(string: url)
-            else {
-                #if DEBUG
-                    print("MiniAppSDK: Invalid URL from manifest.")
-                #endif
-                return
+    func download(with apiClient: MiniAppClient, appId: String, versionId: String, completionHandler: @escaping (Result<URL?, Error>) -> Void) {
+        self.miniAppClient = apiClient
+        self.manifestDownloader.fetchManifest(apiClient: apiClient, appId: appId, versionId: versionId) { (result) in
+            switch result {
+            case .success(let responseData):
+                self.completionHandler = completionHandler
+                self.downloadManifestFiles(with: appId, versionId: versionId, manifestResponse: responseData)
+            case .failure(let error):
+                return completionHandler(.failure(error))
             }
+        }
+    }
 
-            urlToDirectoryMap[url] = pathDirectory.appendingPathComponent(fileDirectory)
+    private func downloadManifestFiles(with appId: String, versionId: String, manifestResponse: ManifestResponse) {
+        guard let miniAppStoragePath = FileManager.getMiniAppDirectory(with: appId, and: versionId) else {
+            return
+        }
+        miniAppDirectory = miniAppStoragePath
+        downloadMiniApp(manifestResponse.files, to: miniAppStoragePath)
+    }
 
+    private func downloadMiniApp(_ urls: [String], to miniAppPath: URL) {
+        self.miniAppClient?.delegate = self
+        for url in urls {
+            guard let fileDirectory = UrlParser.parseForFileDirectory(with: url) else {
+                self.completionHandler?(.failure(NSError.downloadingFailed()))
+                return
+            }
+            urlToDirectoryMap[url] = miniAppPath.appendingPathComponent(fileDirectory)
             queue.addOperation {
-                let task = self.session.downloadTask(with: downloadUrl)
-                task.resume()
+                self.miniAppClient?.download(url: url)
             }
         }
     }
 }
 
-/**
- * URLSessionDownloadDelegate methods for MiniAppDownloader.
- */
-extension MiniAppDownloader: URLSessionDownloadDelegate {
-    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
-        guard let sourceURL = downloadTask.originalRequest?.url else {
+extension MiniAppDownloader: MiniAppDownloaderProtocol {
+
+    func downloadedFileData(downloadTask: URLSessionDownloadTask, location: URL) {
```

Comment:
This seems to just need a `URL` passed in?

Reply:
Done